### PR TITLE
 Add user information export feature

### PR DESCRIPTION
**Pull request by madurangasiriwardena**
### Proposed changes in this pull request

- Add implementation of user information export

### Follow up actions

- Merge https://github.com/wso2/carbon-identity-framework/pull/1309

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.user.endpoint/src/main/java/org/wso2/carbon/identity/user/endpoint/Util/Utils.java [hunk 1]
```diff
@@ -190,4 +192,16 @@ public static User getUser(SelfRegistrationUserDTO userDTO) {
...
 
+    public static UserInformationService getUserAttributesExporter() {
+
+        return (UserInformationService) PrivilegedCarbonContext.getThreadLocalCarbonContext()
+                .getOSGiService(UserInformationService.class, null);
+    }
+
+    public static RealmService getRealmService() {
+
+        return (RealmService) PrivilegedCarbonContext.getThreadLocalCarbonContext()
```
Comment (omindu) [inline components/org.wso2.carbon.identity.user.endpoint/src/main/java/org/wso2/carbon/identity/user/endpoint/Util/Utils.java]: Do a null check

File: components/org.wso2.carbon.identity.user.endpoint/src/main/java/org/wso2/carbon/identity/user/endpoint/impl/PiInfoApiServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.user.core.util.UserCoreUtil;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import java.io.UnsupportedEncodingException;
+import java.util.Base64;
+import java.util.Map;
+import javax.ws.rs.core.Response;
+
+public class PiInfoApiServiceImpl extends PiInfoApiService {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.user.endpoint/src/main/java/org/wso2/carbon/identity/user/endpoint/impl/PiInfoApiServiceImpl.java]: Class level comment

File: components/org.wso2.carbon.identity.user.endpoint/src/main/java/org/wso2/carbon/identity/user/endpoint/impl/PiInfoApiServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Override
+    public Response getUserById(String userId) {
+
+        String usernameFromRequest;
+        try {
+            usernameFromRequest = new String(Base64.getUrlDecoder().decode(userId.getBytes("UTF-8")), "UTF-8");
```
Comment (omindu) [inline components/org.wso2.carbon.identity.user.endpoint/src/main/java/org/wso2/carbon/identity/user/endpoint/impl/PiInfoApiServiceImpl.java]: Use `StandardCharset`

File: components/org.wso2.carbon.identity.user.endpoint/src/main/resources/api.identity.user.yaml [hunk 4]
```diff
@@ -160,24 +178,75 @@ paths:
...
              $ref: '#/definitions/Error'
       tags:
-        - Self Register  
+        - Self Register
+
+  /pi-info:
+    get:
+      tags:
+      - User Export
+      summary: Search user by username
```
Comment (omindu) [inline components/org.wso2.carbon.identity.user.endpoint/src/main/resources/api.identity.user.yaml]: Better to explain the usage of the API properly

File: components/org.wso2.carbon.identity.user.endpoint/src/main/resources/api.identity.user.yaml [hunk 5]
```diff
@@ -283,3 +352,247 @@ definitions:
...
           $ref: '#/definitions/Property'
+
+#-----------------------------------------------------
+# The Basic User Profile Information
+#----------------------------------------------------
+
+  BasicProfile:
+    type: object
+    properties:
+      http://wso2.org/claims/userid:
```
Comment (omindu) [inline components/org.wso2.carbon.identity.user.endpoint/src/main/resources/api.identity.user.yaml]: Define this as a map

File: components/org.wso2.carbon.identity.user.export.core/pom.xml [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                            javax.servlet.http; version="${imp.pkg.version.javax.servlet}",
+                            org.wso2.carbon.user.api; version="${carbon.user.api.imp.pkg.version.range}",
+                            org.wso2.carbon.user.core.*; version="${carbon.kernel.package.import.version.range}",
+                            org.wso2.carbon.context; version="${carbon.kernel.package.import.version.range}",
+                        </Import-Package>
+                        <Export-Package>
+                            !org.wso2.carbon.identity.user.export.core.internal,
+                            org.wso2.carbon.identity.user.export.core.*;version="${identity.governance.exp.pkg.version}"
+                        </Export-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (omindu) [inline components/org.wso2.carbon.identity.user.export.core/pom.xml]: ahem! ;)

File: components/org.wso2.carbon.identity.user.export.core/src/main/java/org/wso2/carbon/identity/user/export/core/internal/service/impl/BasicUserInformationProvider.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        try {
+            userClaimValues = getUserStoreManager(tenantId, userStoreDomain).getUserClaimValues(username, null);
+        } catch (UserStoreException e) {
+            throw new UserExportException("Error while retrieving the user information.", e);
+        }
+
+        if (userClaimValues != null) {
+            Map<String, String> attributes = new HashMap<>();
+
+            for (Claim claim : userClaimValues) {
```
Comment (omindu) [inline components/org.wso2.carbon.identity.user.export.core/src/main/java/org/wso2/carbon/identity/user/export/core/internal/service/impl/BasicUserInformationProvider.java]: Shall we use streams.

**Comment by madurangasiriwardena** [comment]
@omindu @IsuraD @ayshsandu @darshanasbg

File: components/org.wso2.carbon.identity.user.endpoint/src/test/resources/repository/conf/carbon.xml [hunk 8]
```diff
@@ -0,0 +1,684 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+<!--
+ Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.user.endpoint/src/test/resources/repository/conf/carbon.xml]: 2018?

File: components/org.wso2.carbon.identity.user.export.core/src/main/java/org/wso2/carbon/identity/user/export/core/internal/service/impl/ConsentInformationProvider.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    @Override
+    public UserInformationDTO getRetainedUserInformation(String username, String userStoreDomain, int tenantId) throws UserExportException {
+
+        try {
+            List<ConsentReceiptDTO> receipts = new ArrayList<>();
+            int limit = 10;
+            int offset = 0;
+            String tenantDomain = realmService.getTenantManager().getDomain(tenantId);
+
+            List<ReceiptListResponse> receiptListResponses = consentManager.searchReceipts(limit, offset, UserCoreUtil
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.user.export.core/src/main/java/org/wso2/carbon/identity/user/export/core/internal/service/impl/ConsentInformationProvider.java]: can't we simplify this like 



    List<ReceiptListResponse> receiptListResponses
    do {
       receiptListResponses = consentManager.searchReceipts(limit, offset, UserCoreUtil
                      .addDomainToName(username, userStoreDomain), tenantDomain,
              null, ConsentConstants.ACTIVE_STATE);
      for (ReceiptListResponse receiptListResponse : receiptListResponses) {
        String receiptId = receiptListResponse.getConsentReceiptId();

        Receipt receipt = consentManager.getReceipt(receiptId);
        receipts.add(Utils.getConsentReceiptDTO(receipt));
      }
      offset += limit;
    }
    while (receiptListResponses != null && receiptListResponses.size() != 0);

**Pull request merged**